src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/validation/HaFlowValidationFsm.java

```diff
@@ -0,0 +1,340 @@
+            getCarrier().sendSpeakerRequest(flowId,
+                    new DumpGroupsForFlowHsRequest(protectedFlowPath.getYPointSwitchId()));
+        });
+
+
+    }
+
+    protected void receivedRules(State from, State to,
+                                 Event event, Object context) {
+        FlowDumpResponse flowDumpResponse = (FlowDumpResponse) context;
+        log.info("The Switch rules received for the switch {}",
+                flowDumpResponse.getFlowSpeakerData().stream().findFirst().map(e -> e.getSwitchId().toString())
+                        .orElse("No switch Id"));
+        receivedRules.add(flowDumpResponse);
+        awaitingRulesSize--;
+        checkOfCompleteDataCollection();
+    }
+
+    protected void receivedMeters(State from, State to,
+                                  Event event, Object context) {
+        MeterDumpResponse meterDumpResponse = (MeterDumpResponse) context;
+        log.info("The Switch meters received for the switch {}",
+                meterDumpResponse.getMeterSpeakerData().stream().findFirst().map(e -> e.getSwitchId().toString())
+                        .orElse("No switch Id"));
+        receivedMeters.add(meterDumpResponse);
+        awaitingMetersSize--;
+        checkOfCompleteDataCollection();
+    }
+
+    protected void receivedGroups(State from, State to,
+                                  Event event, Object context) {
+        GroupDumpResponse groupDumpResponse = (GroupDumpResponse) context;
+        log.info("The Switch groups received for the switch {}",
+                groupDumpResponse.getGroupSpeakerData().stream().findFirst().map(e -> e.getSwitchId().toString())
+                        .orElse("No switch Id"));
+        receivedGroups.add(groupDumpResponse);
+        awaitingGroupsSize--;
+        checkOfCompleteDataCollection();
+    }
+
+    private void checkOfCompleteDataCollection() {
+        if (awaitingRulesSize == 0 && awaitingMetersSize == 0 && awaitingGroupsSize == 0) {
+            fire(DATA_READY);
+        }
+    }
+
+    protected void validateFlow(State from, State to, Event event, Object context) {
+        try {
+            response = service.validateFlow(flowId, receivedRules, receivedMeters, receivedGroups, switchIdSet);
+        } catch (FlowNotFoundException e) {
+            log.error("The haFlow {} not found during the flow validation", flowId, e);
+            sendException(e.getMessage(), "The HaFlow validation operation in the HaFlowValidationFsm",
+                    ErrorType.NOT_FOUND);
+        } catch (SwitchNotFoundException e) {
+            log.error(e.getMessage(), e);
+            sendException(e.getMessage(), "The haFlow validation operation in the HaFlowValidationFsm",
+                    ErrorType.NOT_FOUND);
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+            sendException(e.getMessage(), "The HaFlow validation operation in the HaFlowValidationFsm",
```

Comment:
I think you can use constant for this message because you use same message 3 times

Reply:
good suggestion, fixed